sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -1575,4 +1575,51 @@ class AdaptiveQueryExecSuite
       checkNoCoalescePartitions(df.sort($"key"), ENSURE_REQUIREMENTS)
     }
   }
+
+  test("SPARK-34980: Support coalesce partition through union") {
+    def checkResultPartition(
+      df: Dataset[Row], shuffleReaderNumber: Int, partitionNumber: Int): Unit = {
+      df.collect()
+      assert(
+        collect(df.queryExecution.executedPlan) {
+          case s: CustomShuffleReaderExec => s
+        }.size === shuffleReaderNumber
+      )
+      assert(df.rdd.partitions.length === partitionNumber)
+    }
+
+    withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true",
+      SQLConf.COALESCE_PARTITIONS_ENABLED.key -> "true",
+      SQLConf.ADVISORY_PARTITION_SIZE_IN_BYTES.key -> "1048576",
+      SQLConf.COALESCE_PARTITIONS_MIN_PARTITION_NUM.key -> "1",
+      SQLConf.SHUFFLE_PARTITIONS.key -> "10") {
+      val df1 = spark.sparkContext.parallelize(
+        (1 to 10).map(i => TestData(i, i.toString)), 2).toDF()
+      val df2 = spark.sparkContext.parallelize(
+        (1 to 10).map(i => TestData(i, i.toString)), 4).toDF()
+
+      // positive test
+      checkResultPartition(
+        df1.groupBy("key").count().unionAll(df2),
+        1,
+        1 + 4)
+
+      checkResultPartition(
+        df1.groupBy("key").count().unionAll(df2).unionAll(df1),
+        1,
+        1 + 4 + 2)
+
+      checkResultPartition(
+        df1.groupBy("key").count().unionAll(df2).unionAll(df1.groupBy("key").count()),
+        2,
+        1 + 4 + 1)
```

Comment:
IIUC these physical plans have a single union exec because of `CombineUnions`? Could you add tests for physical plans having multiple union execs?

Reply:
+1

Reply:
I considered about it.. OK will add it later.